mods/mc_core/freeze.lua

```diff
@@ -0,0 +1,91 @@
+
+-- Frozen players
+minetest.register_entity("mc_core:frozen_player", {
+	-- This entity needs to be visible otherwise the frozen player won't be visible.
+	initial_properties = {
+		visual = "sprite",
+		visual_size = { x = 0, y = 0 },
+		textures = {"blank.png"},
+		physical = false, -- Disable collision
+		pointable = false, -- Disable selection box
+		makes_footstep_sound = false,
+	},
+
+	on_step = function(self, dtime)
+		local player = self.pname and minetest.get_player_by_name(self.pname)
+		if not player or not mc_core.is_frozen(player) then
+			self.object:remove()
+			return
+		end
+	end,
+
+	set_frozen_player = function(self, player)
+		self.pname = player:get_player_name()
+		player:set_attach(self.object, "", {x = 0, y = 0, z = 0 }, { x = 0, y = 0, z = 0 })
+	end,
+})
+
+local function freeze_player(player)
+    local parent = player:get_attach()
+    if parent and parent:get_luaentity() and parent:get_luaentity().set_frozen_player then
+        return
+    end
+    local obj = minetest.add_entity(player:get_pos(), "mc_core:frozen_player")
+    obj:get_luaentity():set_frozen_player(player)
+end
+
+local function unfreeze_player(player)
+    local pname = player:get_player_name()
+    local objects = minetest.get_objects_inside_radius(player:get_pos(), 2)
+    for _,obj in pairs(objects) do
+        local entity = obj:get_luaentity()
+        if entity and entity.set_frozen_player and entity.pname == pname then
+            obj:remove()
+        end
+    end
+end
+
+---@public
+---Freezes a player, preventing them from moving on their own
+---@param player ObjectRef to freeze
+function mc_core.freeze(player)
+    local pmeta = player and player:is_player() and player:get_meta()
+    if pmeta then
+        freeze_player(player)
+	    pmeta:set_string("mc_core:frozen", minetest.serialize(true))
+    end
+end
+
+---@public
+---Unfreezes a player, allowing them to move on their own
+---@param player ObjectRef to unfreeze
+function mc_core.unfreeze(player)
+    local pmeta = player and player:is_player() and player:get_meta()
+    if pmeta then
+        unfreeze_player(player)
+	    pmeta:set_string("mc_core:frozen", "")
+    end
+end
+
+---@public
+---Returns true if the player is frozen, false otherwise
+---@param player ObjectRef to check
+function mc_core.is_frozen(player)
+    local pmeta = player and player:is_player() and player:get_meta()
+	return pmeta and minetest.deserialize(pmeta:get("mc_core:frozen")) or false
+end
+
+---@public
+---Runs a function on player, temporarily unfreezing them while the function runs if they are frozen
+---Arguments after the function definition will be passed into in, similar to minetest.after 
+---The player argument will not be passed into func by default, so it should be added to the argument list in the appropriate position
+---@param player ObjectRef to treat as unfrozen
+---@param func Function to run
+function mc_core.run_unfrozen(player, func, ...)
```

Comment:
Can this not be a more general `mc_core.pass_function` function? Just thinking that the naming and use is not intuitive. For example, can we not just have another `mc_teacher.run_unfreeze` function that is passed to this `mc_core.pass_function`? This would make the use of this "core" function much more flexible.

Reply:
Theoretically, this could be made more general in the future, but I think that may be a bit too much of an abstraction.
The main reason that this is its own function is because it's using the local `freeze_player` and `unfreeze_player` functions, which don't alter a player's `mc_core:frozen` metadata attribute (which is used in `mc_core.is_frozen`).
If someone wanted to do something similar with other functions, they could use a check before the inner function (`func`), and call the appropriate outer functions before and after the inner function if that check passes. Abstracting that out into a theoretical `mc_core.pass_function` function would likely create a clunky interface, since `mc_core.pass_function` would have to handle all of the involved functions and their arguments (which can vary between functions)

Reply:
> Just thinking that the naming and use is not intuitive.

Noted, will try to improve that in the future